### ISSUE TITLE
Possibility to conditionally deactivate CS8509 (not exhaustive switch)

### ISSUE THREAD

**Issue by hugener** [dotnet/roslyn#56936]
**Brief description:**

I'm playing with discriminated unions in C#. My goal is to have something that will fill-in until DUs are natively supported.
So far I have a package analyzing switch statements and expressions. ([Code](https://github.com/sundews/Sundew.DiscriminatedUnions) and [NuGet package](https://www.nuget.org/packages/Sundew.DiscriminatedUnions/0.1.0)) 

The analyzer checks whether switches happens on a class/records marked with an DiscriminatedUnion attribute and does its checks to see if all cases are handled. Additionally, it's analyzing the cases to ensure that they are a "closed" inheritance hierarchy.
In switch expressions, I get a CS8509 warning stating that the switch is not exhaustive, although it should be safe to ignore the warning in this case. Is there a way to indicate to the warning that the switch is handled by another analyzer, or could something like this be added?

Similarly, switch statement resembling switch expressions (with return) receive an CS0161 not all code path return a value.

For now I added another diagnostic that the default case should throw an UnreachableCaseException, but I would really like to get rid of that.

**Languages applicable:**

C# for now, should be easy to support VB.

**Code example that the analyzer should report:**

CS8509 is reported and SDU9999 ('Result?' is a discriminated union and should throw UnreachableCaseException in default case)
```csharp
Result? result = Compute("Success");
var message = result switch
{
    Result.Success => "Great",
    Result.Warning { Message: "Tough warning" } => "Not good",
    Result.Warning warning => warning.Message,
    Result.Error { Code: > 70 } error => $"High Error code: {error.Code}",
    Result.Error error => $"Error code: {error.Code}",
    null => "Nothing",
};
```
SDU9999 is only there to force adding the default case like so:
```csharp
_ => throw new UnreachableCaseException(typeof(Result)),
```

**Comment by RikkiGibson** [comment]
Are you looking for a [diagnostic suppressor](https://gist.github.com/mavasani/fcac17a9581b5c54cef8a689eeec954a)?

**Comment by hugener** [comment]
@RikkiGibson thanks, that's solves it for CS8509.

**Comment by RikkiGibson** [comment]
It's possible that some bug or configuration issue prevents the suppressor from functioning in the IDE. If you have a minimal reproducer it would be great if you could file a new issue with it.

**Comment by hugener** [comment]
Hi @RikkiGibson 

I reproduced two issues here:

1. The red squiggly line (VS problem).
2. Codeanalysis appears to report false FlowState, when a switch handles null (CodeAnalysis problem).

[DUTester.zip](https://github.com/dotnet/roslyn/files/7303341/DUTester.zip)

```csharp
var option = Compute(args.FirstOrDefault() ?? "test");
var message = option switch // Red squiggly lines (The 'switch' expression does not handle all...), no compiler error, CS8509 suppressed by analyzer
{
    Option<int>.Some some => $"The value is {some.Value}",
    Option<int>.None => "There is no value",
    null => "Null", // option appears to be null annotated because of this null case.
                    // Does that make sense?
                    // Compute returns a non-nullable Option<int>, so the analyzer should have reported the null case as unreachable.
                    // It does not because CodeAnalysis Nullability on the TypeInfo reports its FlowState as "MaybeNull". Removing the null case changes the FlowState to NotNull.
                    // Reproduced here: https://github.com/sundews/Sundew.DiscriminatedUnions/blob/main/Source/Sundew.DiscriminatedUnions.Test/GenericSwitchExpressionAnalyzerTests.cs
                    // Test: Given_SwitchExpressionInEnabledNullableContext_When_ValueIsNotNullAndAllCasesAndNullCaseAreHandled_Then_HasUnreachableNullCaseIsReported2
};
```

Should I create separate issues for them both in this repository?
